pkg/rc/replication_controller.go

```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+		if err != nil {
+			return err
+		}
+
+		err := rc.scheduleWithSession(newNode)
+		if err != nil {
+			// Different RC may have already taken over the new node, abort transfer
+			rc.quitTransfer()
+			return err
+		}
 	} else {
-		err = rc.transferPetNodes(ineligibleCurrent)
+		newNode = status.NodeTransfer.NewNode
 	}
-	if err != nil {
-		return util.Errorf("Unable to transfer nodes: %s", err)
+
+	if rc.isWatchingHealth {
+		return nil
+	}
+
+	rc.nodeTransfer = nodeTransfer{
+		newNode: newNode,
+		quit:    make(chan struct{}),
 	}
 
+	go rc.watchHealth()
+	rc.isWatchingHealth = true
+
 	return nil
 }
 
-func (rc *replicationController) transferCattleNodes(ineligibleCurrent []types.NodeName) error {
-	if len(ineligibleCurrent) < 1 {
-		return util.Errorf("Need at least one ineligible node to transfer from")
+func (rc *replicationController) updateAllocations(ineligible []types.NodeName) (types.NodeName, error) {
+	if len(ineligible) < 1 {
+		return "", util.Errorf("Need at least one ineligible node to transfer from, had 0")
 	}
 
-	status, queryMeta, statusErr := rc.rcStatusStore.Get(rc.ID())
-	if statusErr != nil && !statusstore.IsNoStatus(statusErr) {
-		return util.Errorf("Unable to get rc status: %s", statusErr)
-	}
+	rc.mu.Lock()
+	man := rc.Manifest
+	sel := rc.NodeSelector
+	rc.mu.Unlock()
 
-	if status.NodeTransfer != nil {
-		// Node transfer already in progress
-		return nil
+	oldNode := ineligible[0]
+	err := rc.scheduler.DeallocateNodes(sel, []types.NodeName{oldNode})
+	if err != nil {
+		return "", util.Errorf("Could not deallocate from %s: %s", oldNode, err)
 	}
 
 	nodesRequested := 1 // We only support one node transfer at a time right now
-	rc.mu.Lock()
-	newNodes, err := rc.scheduler.AllocateNodes(rc.Manifest, rc.NodeSelector, nodesRequested)
-	rc.mu.Unlock()
+	newNodes, err := rc.scheduler.AllocateNodes(man, sel, nodesRequested)
 	if err != nil || len(newNodes) < 1 {
 		errMsg := fmt.Sprintf("Unable to allocate nodes over grpc: %s", err)
 		err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
 		if err != nil {
 			rc.logger.WithError(err).Errorln("Unable to send alert")
 		}
 
-		return util.Errorf(errMsg)
+		return "", util.Errorf(errMsg)
 	}
 
-	var modifyIndex uint64
-	if statusErr != nil && statusstore.IsNoStatus(statusErr) {
-		modifyIndex = 0
-	} else {
-		modifyIndex = queryMeta.LastIndex
-	}
+	newNode := newNodes[0]
 
-	status.NodeTransfer = &rcstatus.NodeTransfer{
-		// Again, we only support one node transfer at a time so just take the first old and new
-		OldNode: ineligibleCurrent[0],
-		NewNode: newNodes[0],
+	status := rcstatus.Status{
+		NodeTransfer: &rcstatus.NodeTransfer{
+			OldNode: oldNode,
+			NewNode: newNode,
+		},
 	}
+
 	writeCtx, writeCancel := transaction.New(context.Background())
 	defer writeCancel()
-	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), modifyIndex, status)
+	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), 0, status)
```

Comment:
I think we shouldn't assume that the status doesn't exist because we might want to track other information in the status key someday (for example maybe we want to store the full list of current nodes to reduce the cost of looking that up)

Reply:
We could consider using a different status key in consul for that though to make coordination here easier

Reply:
I think ultimately my decision to do it this way (can assume non-existence of status; always delete the status when transfer is done) was because of frustration with `CASTxn`. It seemed like a lot do a `Get`, `CASTxn`, and then a `transaction.Commit` just to try and eliminate history of the node transfer. Its particularly verbose since each of those can return `error`. I preferred just being able to delete and use 0 as the index. That is my argument in favor of having a separate status. That couldv'e been laziness and I understand that might just defer complication by requiring a separate status down the road.

If we keep this as the only rc status, should I try to identify where a `Set` would work, or could a `CAS` without a transaction ever be appropriate? Or is `CASTxn` the best route?

Reply:
as a general statement, I feel like using transactional operations even if they're the only operation is still good because it makes it easier to add more operations to the transaction later on if it makes sense.

I think the argument for keeping separate status records is pretty strong, we can accomplish this by using different namespaces when configuring the status store. We should probably make the namespace string mention node transfers